### DISCUSSION TITLE
Compatibility Mode? How to help people coming from Py.Processing

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#12] [Ideas]
Hi, @hx2A 

First I would like to express my feelings of joy for the existence of this work of yours, I hope many people may join you to help and make this something even more wonderful everyone can use.

Do you think we could have some kind of "Compatibility Mode" to help people migrate examples from Processing Python mode?

I have tons of examples and didactic materials with the the traditional Processing naming conventions like `noFill()` and `mousePressed` etc.

**Comment by hx2A** [comment]
@villares , thank you for your kind words! It makes me very happy to hear how py5 brings you joy. I do hope many join together to build a community of people who make this something even better. There's still a lot to be done, particularly around documentation and example code. You are welcome to join this emerging community!

> Do you think we could have some kind of "Compatibility Mode" to help people migrate examples from Processing Python mode?

I have written code to covert Processing (Java) to py5 (Python), and used it for the example code you see on most pages in the reference documentation. It is not possible to do a perfect conversion but I could get it to work 90% of the time. I only had to use that once and I just manually fixed the errors. The basic idea behind that was to tokenize the code and then convert camel case to snake case. However, note that some Processing functions such as `map` and `filter` are called `remap` and `apply_filter` in py5 to avoid conflicts with Python builtins.

Converting processing.py code to py5 would be much easier because they both use python syntax already. My suggestion would be to use Python's `tokenize` library to tokenize processing.py code, and then read through the tokens to identify what needs to be changed. You can build a mapping of Processing names to py5 names using these data tables:

https://github.com/hx2A/py5generator/tree/main/py5_resources/data

I don't think a perfect translation will be possible but something that does 98% of the work is very achievable. Perhaps we could build a utility function and add it to py5_tools, so it becomes a part of the library? That would be a great contribution.

**Reply by villares** [reply]
I was thinking initially about having legacy names that could act as aliases to the new Python-style names.

It is possible to make `map()`, `set()` and a few other conflicting names work differently according to the nature of the arguments, like jdf did, and then Bernardo... Another challenge is to have a `mousePressed`/`keyPressed` work both as a boolean global and a user defined callable... There are days I feel like this would be very important to bring in Processing people and have people benefit from the huge stash of examples, but on other days I feel like we should move into the future a bit like p5js did. I really oscillate between those two moods.

As for converting from Java I have written a few words to get my students port stuff into Python mode, and that seems to have helped them a bit: https://abav.lugaralgum.com/py.processing-play/java_to_python.html, and I have tried my hand on some pretty crappy automation... you documentation looks great!

**Reply by hx2A** [reply]
What do you mean by legacy names?

> It is possible to make `map()`, `set()` and a few other conflicting names work differently according to the nature of the arguments,

Does Jython support multiple dispatch like Java does? Since Python does not, I can't really have map and set work differently according to the arguments. Or perhaps I could replace the builtins and do something complicated that analyzes the arguments, but I think it is much more straightforward and clear to have builtins be builtins and and for everything else to avoid name conflicts. I think it is better for beginners because beginners need to know what those builtins are.

> Another challenge is to have a mousePressed/keyPressed work both as a boolean global and a user defined callable

In py5, I changed those to `is_mouse_pressed` and `is_key_pressed` to avoid the conflict with the user defined callables. There's also `get_frame_rate()` to get avoid a conflict with `frame_rate(float  rate)`. I can't have a variable and a method have the same name. JPype and pyjnius both prioritize methods over fields.

Is this the code you were referring to?

https://github.com/villares/py.processing-play

Let me take a stab at writing a converter.

Your documentation on converting Processing to Python is very thorough!

**Reply by villares** [reply]
Yes it is possible with some shenanigans to have some "multiple dispatch" behavior in Python. It's been done twice, by JDF and by Bernardo, shadowing built-ins, I know it would be a hard decision, but it could be turned on and off perhaps... I don't believe it creates a problem for beginners, but I don't know. I don't want to create tension or polemic!

By legacy names I mean the Py.Processing/Processing function names we've been using for years, `noFill()`, `lerpColor()` everything. Compatibility mode would be a drop in replacement to Py.Processing (in my dreams).

I'm curious, why not rename `settings()` to `setup()`? 
![image](https://user-images.githubusercontent.com/3694604/120085908-90bdc680-c0b2-11eb-83f7-36d407f0c654.png)

**Reply by hx2A** [reply]
This discussion has done a lot to help me think through some of the documentation that I need to write. Thanks!

I realized last night that I should start by stating what the purpose / goal of py5 is. Referencing jdf's comment in another issue thread:

> The purpose of my project was to make it possible to do Processing programming with Python *syntax*, not to integrate Processing with Python.

My goal is to create a new version of Processing that fits into the larger Python ecosystem. Python syntax is a byproduct of that, but it isn't the core goal. This also isn't an attempt to replace or compete with any other version of Processing. Py5 isn't and can't be a drop-in replacement for processing.py, even without the camel case / snake case differences. I made different design choices in pursuit of a different goal, resulting in different libraries that work differently.

The reason why I used the snake case naming convention is because that is what people in the Python world usually expect, and I anticipate users will code py5 along side other libraries that also use the same convention. In my view, mixing snake case with camel case is problematic. 

I do hear what you are saying though: you've put a lot of time and effort into developing educational content for processing.py, and if you wanted to convert it to py5, you don't want to spend time making tedious camel case => snake case changes.

To help, I took a stab at writing a translator that automates the tedious editing. Look at the code I added to the new translators branch. I tried running it on your repo; most of the code either worked right away or with minor changes. I estimate I could get most of that code using py5 in under an hour. 

Here are some of the things I saw that I would be doing in that hour:

* py5 works with other Processing Java libraries like peasycam but requires different code to add/utilize the libraries
* PVector and PMatrix are gone, but you can use numpy arrays instead
* Processing provides support for threading, but those functions can't work with py5. py5 also offers support for threading, but it works differently
* 2.7 to 3 issues
* Instead of using Processing's libraries for an Arduino or a Webcam, consider using the Python libraries pyserial or opencv.

Also keep in mind that py5 might offer different or better ways of doing things. For example, the `np_pixels` array lets you work with pixels in a 3 dimensional array much like every other Python image library. The 1 dimensional `pixels` array is still available, but why use it if you can use `np_pixels` instead?

**Reply by hx2A** [reply]
> I'm curious, why not rename `settings()` to `setup()`?

Because `settings()` and `setup()` are two different functions. Just like in Processing, calls to `size` and some other initialization type calls need to get in `settings()`. `setup()` is for things that are run once.

The PDE does this trick where you can put the contents of `settings()` into `setup()` and it will separate it for you before running the code. That only works in the PDE though; if you write your own Java classes in another editor, you will need to call `settings()` yourself.

py5 doesn't currently let you combine the two functions. I attempted to support that several times before but could not get it work reliably or without negative side effects. I might get that working someday, but don't bet on it.

**Reply by villares** [reply]
I see, it is another layer of magic the PDE adds to simplify things to the end-user. 

You should not bet on solving everything yourself either! Let's bring in more people to help, like @berinhard and @jsbueno ...

**Reply by hx2A** [reply]
I would appreciate the help! Right now I need people who can fix the OSX platform issues. That's the biggest thorn in my side.

**Reply by hx2A** [reply]
I have finished the code to split a `setup()` function into `settings()` and `setup()`. It works very well, and even works correctly when there are variables in the fake settings() method:

```
w, h = 200, 200

def setup():
    size(w, h, P2D)
```
Coding this was pretty easy once I found the right approach. Thanks for encouraging me to take another look at this.

This feature works in imported mode and module mode. It does not work for Sketches implemented in class mode.

I will update the documentation and example code to reflect this change. The next release will be in early July, most likely.

**Reply by hx2A** [reply]
Also, I wanted to ask you: what languages do you speak besides English and Portuguese?

**Reply by villares** [reply]
Thank you @hx2A, wonderful stuff!

Why do you ask? I can speak a bit of French (but I wouldn't dare write any), and I can understand some Spanish.
Maybe you saw me interacting with the brilliant Japanese Processing community on Twitter, but I use translation tools to read their tweets and then exchange messages in English.

**Reply by hx2A** [reply]
You'll have to tell me more about these translation tools for Twitter. I'm not on Twitter now but am starting to believe I should be using it, if I can learn how to do so effectively. You make it sound like a worthwhile tool for communicating with people and not just a mechanism for people to spew bile.

I have a half-baked idea for the future of py5 that involves multiple (spoken) languages. I'll ask you more questions about it later!

**Reply by villares** [reply]
there is a translate button https://twitter.com/ryotakob/status/1406571432146997248
![image](https://user-images.githubusercontent.com/3694604/123010177-7070e700-d394-11eb-8985-0393fbb851e3.png)

**Reply by hx2A** [reply]
@villares Hmmm, I don't see that translate button. Is this a browser extension you've added?

**Reply by villares** [reply]
No, I'm just using Chromium (Chrome without extra Google telemetry) and Firefox also works, a small "translate" blue link appears, but you have to click on the tweet l suppose (it enters a thread reading mode I guess). 
![image](https://user-images.githubusercontent.com/3694604/123132608-4c111b00-d425-11eb-8951-cff62e420ba7.png)

**Reply by hx2A** [reply]
I see. It probably only appears for people who are logged in. In any case, I intend to make use of this feature when I join.

**Comment by villares** [comment]
Oh,  you can make a class that implements callabe objects with `__call__()` that also have a `__bool__()`, it is doable :D

**Reply by villares** [reply]
Oh, don't bother with a converter for my stuff! Actually I have quite a few projects done in Processing Python mode, three years of daily sketches etc...

**Reply by hx2A** [reply]
Already done! I needed to invest time in translators anyway. Eventually I will need to convert imported mode <=> module mode.

**Comment by villares** [comment]
In my opinion, the grand strategy to bring everything to the global namespace, and maybe using `import *`, is to use `run_sketch()` or something to monkey patch, kind of "decorate" `draw()`, so it will update all the globals (like `key`, `mouse_x`, `mouse_y` & etc).

**Reply by hx2A** [reply]
Imported mode uses AST transformations to get dynamic variables like `mouse_x` and `mouse_y` to work. It's simple, works reliably, and has no performance side effects.

Keep in mind py5 also has module mode and class mode. The library supports all 3 at the same time. It is interesting to me that you have gravitated towards imported mode. It is the most similar to processing.py?

**Reply by villares** [reply]
Yes I believe it is crucial not to need `py5.` for teaching children and beginners, the less elements you need, the less cognitive load you waste, when they are already overwhelmed by so many new concepts. Like, when they are learning indentation and variables... https://github.com/villares/material-aulas/blob/main/Processing-Python/indentacao.md

I've been reading some computing education research that matches closely what people from The Carpentries has been practicing: http://teachtogether.tech

But then again, we want ["low floor"  and "high ceiling"](https://arts.ucla.edu/single/a-modern-prometheus/#!), right?

**Reply by villares** [reply]
Think of how Python itself removes clutter, typing `py5.` for every function and constant is like having to add `;` at every line, perhaps worse for those who don't even know why it is there.

**Reply by hx2A** [reply]
What age group do you teach?

> Yes I believe it is crucial not to need py5. for teaching children and beginners, the less elements you need, the less cognitive load you waste, when they are already overwhelmed by so many new concepts. 

When first started coding py5 a year ago I thought the same thing, and tried to implement py5 using an "imported mode" coding style. None of my ideas worked though, and eventually I settled on what is now known as class mode and module mode. I never gave up on being able to skip prepending everything with `py5.` I stuck with it, and finally got something to work.

Internally py5 is really built around class mode, and module mode is the result of some Python magic. Imported mode is some more Python magic built around module mode.

**Reply by villares** [reply]
Every age group possible, from 10 years old onward. Nowadays more high-schoolers and adults to be fair. 

I understand it is a huge technical challenge... I'm not skilled enough to help on this, but there are many nice people you might want to chat with.

**Reply by hx2A** [reply]
Right now the bottleneck is my time and ability to write documentation. My current goal is to get enough of the right documentation written so that people who know Processing and Python well can get up to speed to start using this and help teach others.

**Reply by villares** [reply]
I have asked a very skilled Python meta-programmer to have a look.
I think removing `py5.` for basic usage and `settings()` will have a big impact for beginners.

**Reply by hx2A** [reply]
> I think removing ... `settings()` will have a big impact for beginners.

Do you say this because beginners who are familiar with Processing or processing.py would expect it to work this way? What about beginners who know nothing about Processing and don't have any expectations about how it works? You have more knowledge and experience about teaching, so I'd like to hear your thoughts on this.

Technically Processing requires `settings()`, but when coding through the PDE, the PDE does some text transformations to the user's code before compiling, and then compiles the transformed code. Those text transformations do have some limitations. For example, this does not work:

```
void setup() {
  if (random(1) > 0.5) {
    size(400, 400);
  } else {
    size(200, 200);
  }
}

void draw() {
  rect(mouseX, mouseY, 10, 10);
}
```

This is the error message I get:

```
When not using the PDE, size() can only be used inside settings().
Remove the size() method from setup(), and add the following:
public void settings() {
  size(400, 400);
}
IllegalStateException: size() cannot be used here, see https://processing.org/reference/size_.html
```

That error message seems a bit confusing to me as it does not fully explain what is wrong with the code. But, people are happy with how this works, so perhaps proper expectations have been set about what can be accomplished with the text transformations done by the PDE.

With py5, I can have the py5 kernel do an AST transformation on a `setup()` function to pull out the `size()` call and construct a synthetic `settings()` function. This is easy to do. The reason why I haven't done it yet is:

 1. the AST transformation can't be done perfectly and will have some limitations, so I will need to make sure it works in a way that conforms to a set of reasonable expectations
 2. any error messages generated by the synthetic function will point to a function that does not actually exist, which might be confusing for beginners

Given that how the PDE handles this has limitations and that people are still happy with how it works suggests to me that I could also come up with something that people will be happy with, even though it will have limitations. The tricky part for me is the part about expectations.

I'll move this up in my priority list, and when I have something to show, I'll let you know and ask you for feedback about if it is suitable for beginners.

**Reply by villares** [reply]
I think being able to work without `.py5` is more important, but regarding `settings()`

Looking from the perspective of a teacher using the Processing IDE (I would like to have something closer to de IDE experience for beginners):
- One can even start drawing without `setup()` and `draw()`, the so called static mode! Very useful for a first contact.
- Then there is this big step to introduce the `setup()`/`draw()` structure, adding in a third requirement adds too much complexity to someone no even sure how a variable works, or how to define their own functions. It's hard for us to remember how it feels like not to know stuff, adding a function name for us means very little cognitive load. For them, less is more in this context.

Naive question:
Why not make a *faux* `size()` call all the settings preparation and then the *true* size function?  The only requirement would then be that you call size() at the beginning of `setup()`, no? Well probably lot's of context I'm missing... anyway, thanks for the patience!

**Reply by hx2A** [reply]
> I think being able to work without `py5.` is more important, but regarding `settings()`

Wait. Isn't that what is being provided by imported mode? With the py5 kernel you do not need to prepend `py5.` everywhere. Does imported mode meet that need?

> One can even start drawing without setup() and draw(), the so called static mode! Very useful for a first contact.

Agreed! That's why I created the cell magics, such as [%%py5draw](http://py5.ixora.io/reference/py5draw/). They should work just fine in imported mode, again without the `py5.` prefix everywhere.

However, feedback I've gotten from another educator is that the IPython magics aren't beginner friendly, because they are like a weird non-python command that is getting added to the notebook. I like them, but they may not be good for everyone. Right now an idea I have is to repackage that code to create a second custom kernel, called "py5bot", that works kind of like drawbot and shoebot. What are your thoughts on this?

> Then there is this big step to introduce the setup()/draw() structure, adding in a third requirement adds too much complexity to someone no even sure how a variable works

Good point. I see what you are saying.

> Why not make a faux `size()` call all the settings preparation and then the *true* size function? The only requirement would then is that you call `size()` at the beginning of `setup()`, no? 

This was a thought provoking question, and thinking about intervening with a fake `size()` method generated some good ideas. If the `size()` method were the only one that mattered an approach like this could work, but then I realized that I also need to think about `fullscreen()`, `no_smooth()`, `smooth()`, and `pixel_density()`, all of which need similar treatment.

Because this topic was on my mind, I spent some time exploring it last night. I came up with a working prototype that can split the user's `setup()` function in half, and have py5 use the first one for `settings()` and the second one for `setup()`. Critically, any error messages from buggy code are exactly correct, pointing to the correct line number and source code as written by the user. Error messages pointing to code not written by the user are obviously confusing and not appropriate for beginners, and unhelpful to everyone else. The prototype uses a totally different approach from anything I thought of before, and only requires about 10 lines of code.

I need to get the mybinder stuff out the door so this new feature won't make it into the next release (a few more days), but I will put it in the one after that, probably due out in early July. I'll need help testing that, and will ask you about that when it is ready.

**Reply by hx2A** [reply]
> Right now an idea I have is to repackage that code to create a second custom kernel, called "py5bot", that works kind of like drawbot and shoebot. 

Today I woke up with an idea for how to build py5bot, and a few hours later, I had a working prototype. After playing around with it for a while I now see the value of it. Educators will be able to illustrate programming concepts that zero in on one concept, without excess visual clutter. For example, if I wanted to explain a for loop to someone I would rather write this:

```
size(200, 200)
for i in range(10):
    rect(i * 20, 100, 5, 5)
```

Contrast that with this:

```
def settings():
    size(200, 200)

def setup():
    for i in range(10):
        rect(i * 20, 100, 5, 5)

run_sketch()
```

And the first bit of code will put the image right in the notebook for you. The second piece of code would require even more code for that.

The tricky part about py5bot (and also the `run_sketch.py` utility that @tabreturn used in his Thonny blog post) is that the error messages for both syntax errors and programming exceptions need to show an error message with code that points to the actual code the user wrote. The challenge is that in py5bot the user's code needs to get transformed to something else to actually run in a sketch, but then any exceptions will be generated by the modified code. Presenting a beginner programmer an error message or stack trace for code they didn't write is not helpful.

Anyhow, py5bot basically works but I have  few things I need to straighten out, and I also want to apply what I've learned to `run_sketch.py`, as its error messages for static sketches are currently flawed. I'll have all of this ready for the next release in a few weeks.

**Reply by tabreturn** [reply]
I think I see what you mean. For example, if I enter the following on line 1 of, say, my Thonny editor:
```
rect(10)
```
Then I get:
```
...
File "demo.py", line 5, in setup
    3    def setup():
    4        rect(10)
...
```
So, instead, this would say something like:
```
...
File "demo.py", line 1, in <module>
    1    rect(10)
...
```
Sounds good!

**Reply by hx2A** [reply]
Right. The latter is better.

Syntax errors are handled differently. In `run_sketch.py` the error message is horrific. This code:

```
rect(10))
```
Gives this error message:

```
 run_sketch /tmp/test.py 
Traceback (most recent call last):
  File "/home/jim/INSTALL/anaconda3/bin/run_sketch", line 8, in <module>
    sys.exit(main())
  File "/home/jim/INSTALL/anaconda3/lib/python3.8/site-packages/py5_tools/tools/run_sketch.py", line 34, in main
    imported.run_code(args.sketch_path, classpath=args.classpath)
  File "/home/jim/INSTALL/anaconda3/lib/python3.8/site-packages/py5_tools/imported.py", line 128, in run_code
    _run_sketch(sketch_path, classpath, exit_if_error)
  File "/home/jim/INSTALL/anaconda3/lib/python3.8/site-packages/py5_tools/imported.py", line 104, in _run_sketch
    sketch_ast = ast.parse(sketch_code, mode='exec')
  File "/home/jim/INSTALL/anaconda3/lib/python3.8/ast.py", line 47, in parse
    return compile(source, filename, mode, flags,
  File "<unknown>", line 3
    rect(10))
            ^
SyntaxError: unmatched ')'
```
Only the piece at the end is helpful.

**Reply by hx2A** [reply]
@tabreturn @villares I am pleased to say I have finished py5bot and have merged it into main. I'm very excited about this and can clearly see the value. Thanks again for the encouragement to work on this feature.

py5bot is a second Jupyter Notebook kernel that executes every cell as its own static mode py5 sketch. It supports the default and opengl renderers.  In addition, the py5 kernel has a new IPython magic called `%%py5bot` that makes just that cell into a py5bot cell. I think that will be useful for learners moving from the py5bot kernel to the py5 kernel.

In addition, I reworked the code so that any error messages point to the line with the problem and there are no Python or Java stack traces showing py5's internal code. For example:

```
size(200, 200)
background(255, 0, 0)
rect("error")
```

gives this error:

```
py5 encountered an error in your code:
    1    
    2    background(255, 0, 0)
--> 3    rect("error")

TypeError: The parameter types (str) are invalid for method rect.
Your parameters must match one of the following signatures:
 * rect(a: float, b: float, c: float, d: float, /) -> None
 * rect(a: float, b: float, c: float, d: float, r: float, /) -> None
 * rect(a: float, b: float, c: float, d: float, tl: float, tr: float, br: float, bl: float, /) -> None
```

Once I got that working I applied what I learned to the `run_sketch.py` utility that @tabreturn used with Thonny. So that has the same clean and beginner appropriate error messages. Also, the other magics like `%%py5draw` have better error messages.

I have one OSX bug I want to fix and then I'll do a release, probably this weekend. I'll add py5bot to the py5examples repo so it is available via binder.

The py5bot kernel plus the settings/setup combo both make this an important release. I've also fixed various bugs and straightened out some things I wanted to fix before sharing it with more people.

I need to update the documentation to add these new features. After the release I'm going to fix bugs as they come in and focus much more on making progress with tutorials and how-to's documentation. I do have more ideas for new features, but I'll leave them alone for now.

Also I created a twitter account for py5 and myself and will start using both when I do the release. I'm new to twitter and don't really know how to use either effectively, so any pointers would be helpful.

**Reply by hx2A** [reply]
And I see py5bot doesn't work so well on Macs for the OpenGL renderers. I know what the problem is though. We'll see if I can fix it.

And referring back to your examples, the static sketches `rect(10)` and `rect(10))` will now yield the following errors:

```
py5 encountered an error in your code:
    rect(10)

TypeError: The parameter types (int) are invalid for method rect.
Your parameters must match one of the following signatures:
 * rect(a: float, b: float, c: float, d: float, /) -> None
 * rect(a: float, b: float, c: float, d: float, r: float, /) -> None
 * rect(a: float, b: float, c: float, d: float, tl: float, tr: float, br: float, bl: float, /) -> None
```

```
There is a problem with your code:
SyntaxError: unmatched ')' (<py5bot>, line 1)
    rect(10))
            ^
```

**Reply by tabreturn** [reply]
Great work, @hx2A! I look forward to testing the new release. I'm keen to get involved in documentation, but this month is a bit manic for me. Please let us know your Twitter account (I don't personally have an account, but I follow users using RSS). @villares could probably give you some pointers.

**Reply by hx2A** [reply]
> I'm keen to get involved in documentation, but this month is a bit manic for me

Same here. I'm also busy with a collaboration I'm doing with an artist for a commissioned work. I can't talk about the details but I am using py5 for it. This library is great for prototyping ideas.

I'm eager to work with you on the documentation when you are ready!

> Great work, @hx2A! I look forward to testing the new release.

Thanks!

Today I discovered that py5bot doesn't work at all on OSX, but I figured out exactly why and I know how to fix it. I'm also now close to solving most of my OSX issues. OSX has been a thorn in my side for months. Rather than do a release this weekend I will finish off those problems so I don't have to think about it anymore.

I created py5coding for Twitter. I'll also have a personal Twitter account but haven't made up my mind what the handle will be.

**Comment by villares** [comment]
I haven't seen any examples on the notebook without the `py5.` prefix, so I think I missed this use, even the page you mentioned about the notebook 'magic' invocation has the prefix http://py5.ixora.io/reference/py5draw/
I'll have to find time to test things out! Thanks, and cheers, and try not to do too much!

**Reply by hx2A** [reply]
Hmmm, it is clear to me that the documentation needs work.

Module mode, class mode, and imported mode are all 3 ways of using py5. Everything you can do in one mode can be done in the other two, with some adjustments to the code. Any time you see code in module mode (with `py5.` prefixes) you can translate it to imported mode by erasing `py5.` everywhere.

I see how the reference documentation is not clear about this. I have an idea to present the example code in module mode and imported mode, with a toggle button to switch between the two. That way it will be more clear and the user an see whichever one is more convenient.

I just did the release, and of course found a small bug right after. :)

**Comment by tabreturn** [comment]
Fantastic project, @hx2A! And this is a great thread. I'll be keeping a close eye on py5, and I'll try to contribute where I can.

**Comment by tabreturn** [comment]
Sorry -- yeah, I deleted it to make a new topic: https://github.com/hx2A/py5generator/discussions/19

**Comment by bryanidem** [comment]
Hi! Fantastic initiative! I'm new working on open source projects, but I'll do my best to collaborate in this project. :)

**Reply by hx2A** [reply]
Excellent! I'm looking forward to working with you.

Open source projects are an excellent opportunity to learn from others. You can interact with and learn from people who are outside the group of people you work with in a professional setting. I've learned a ton this way.

There is a [code of conduct](https://github.com/hx2A/py5generator/blob/main/CODE_OF_CONDUCT.md) for this project and for most other open source projects. Open source works best when people create welcoming environments for others from different backgrounds or programming experiences to ask questions and to make contributions.
